singularity-eos/eos/singularity_eos.cpp

```diff
@@ -343,327 +350,575 @@ int get_sg_eos( // sizing information
+            rho_pte(tid, mp) = npte / spvol_v(i) * frac_mass_v(i, m);
+            vfrac_pte(tid, mp) = 1.0 / npte;
+            temp_pte(tid, mp) = temp_v(i) * ev2k;
+          }
+          Real sie_tot_true{0.0};
+          const int neq = npte;
+          singularity::mix_impl::CacheAccessor cache(&solver_scratch(tid, 0) +
+                                                     neq * (neq + 4) + 2 * npte);
+          if (npte > 1) {
+            // create solver lambda
+            // eos accessor
+            singularity::EOSAccessor_ eos_inx(eos_v, &pte_idxs(tid, 0));
+            PTESolverFixedT<singularity::EOSAccessor_, Real *, Real **> method(
+                npte, eos_inx, 1.0, temp_pte(tid, 0), &rho_pte(tid, 0),
+                &vfrac_pte(tid, 0), &sie_pte(tid, 0), &temp_pte(tid, 0),
+                &press_pte(tid, 0), cache, &solver_scratch(tid, 0));
+            const bool res_{PTESolver(method)};
+            // calculate total internal energy
+            for (int mp = 0; mp < npte; ++mp) {
+              const int m = pte_mats(tid, mp);
+              sie_tot_true += sie_pte(tid, mp) * frac_mass_v(i, m);
+            }
+          } else {
+            // pure cell (nmat = 1)
+            // calculate sie from single eos
+            sie_pte(tid, 0) = eos_v(pte_idxs(tid, 0))
+                                  .InternalEnergyFromDensityTemperature(
+                                      rho_pte(tid, 0), temp_pte(tid, 0), cache[0]);
+            // set total sie to material 0 value
+            sie_tot_true = sie_pte(tid, 0);
+            // set pressure
+            press_pte(tid, 0) = eos_v(pte_idxs(tid, 0))
+                                    .PressureFromDensityTemperature(
+                                        rho_pte(tid, 0), temp_pte(tid, 0), cache[0]);
           }
-          spvol_v(i) = r_is_inp ? spvol_v(i) : vol_sum / mass_sum;
-          sie_v(i) = eng_sum / mass_sum;
-        }
-        // calculate component volumes and energies
-        // figure out participating materials, get everything
-        // set up to feed in to solver
-        int npte = 0;
-        double vsum_nopte = 0.0;
-        double esum_nopte = 0.0;
-        // double rhoavg_pte = 0.0;
-        for (int m = 0; m < nmat; ++m) {
-          const bool something = frac_mass_v(i, m) / mass_sum > min_frac;
-          frac_sie_v(i, m) = sie_v(i) * frac_mass_v(i, m);
-          frac_vol_v(i, m) = spvol_v(i) * frac_mass_v(i, m);
-          if (something) {
-            // vfrac true
-            int midx = eos_offsets_v(m) - 1;
-            pte_mats(tid, npte) = midx;
-            pte_idxs(tid, npte) = m;
-            mass_pte(tid, npte) = frac_mass_v(i, m);
-            sie_pte(tid, npte) = frac_sie_v(i, m) / frac_mass_v(i, m);
-            vfrac_pte(tid, npte) = frac_vol_v(i, m);
-            npte += 1;
+          // since rho_ptes are correct, now do rho-T lookups for remaining quantities
+          // first zero out quantities to be averaged
```

Comment:
As I see it, you're using rho-T lookups to fill out all the quantities no matter what the input conditions were right? Can this be put in a function to avoid code duplication?

Reply:
different values need to be set based on the input condition.

Reply:
done